mergify_engine/worker.py

```diff
@@ -0,0 +1,332 @@
+        self._redis = None
+
+        self.pending_streams = set()
+
+    async def stream_worker_task(self, worker_id):
+        # NOTE(sileht): This task must never fail, we don't want write code to reap/clean/respawn them
+        while self._workers_running.is_set():
+            try:
+                try:
+                    stream_name = await asyncio.wait_for(
+                        self._queue.get(), timeout=self._idle_sleep_time
+                    )
+                except asyncio.TimeoutError:
+                    continue
+
+                LOG.debug("stream worker %d take stream: %s", worker_id, stream_name)
+                self.pending_streams.add(stream_name)
+
+                try:
+                    installation_id = stream_name.split("~")[1]
+                    async with self._installation_locks[installation_id]:
+                        async with get_sources_from_stream(
+                            self._redis, stream_name
+                        ) as sources:
+                            await self._loop.run_in_executor(
+                                self._executor,
+                                functools.partial(run_engine, stream_name, sources),
+                            )
+                finally:
+                    LOG.debug(
+                        "stream worker %d release stream: %s", worker_id, stream_name
+                    )
+                    self.pending_streams.remove(stream_name)
+                    self._queue.task_done()
+            except Exception:
+                LOG.error("stream worker %d fail", worker_id, exc_info=True)
+                return
+
+        LOG.info("stream worker %d exited", worker_id)
+
+    async def push_streams_to_worker(self):
+        self._pusher_tombstone.clear()
+        self._pusher_running.set()
+
+        streams = []
+        locked_streams = set()
+        while self._pusher_running.is_set():
+            # The queue is full, put back locked_streams, we can't known which
+            # installation will be unlocked, so unlock all of them, those still locked
+            # will return to the `locked_streams` anyways.
+            if self._queue.full() and locked_streams:
+                streams = list(locked_streams) + streams
+                locked_streams = set()
+
+            # If we have nothing to do pickup new streams, the list may contains
+            # stream we are already handling so we filter them out
+            if not streams:
+                # NOTE(sileht): Only return streams with a date in the past from the oldest to the newest
+                # TODO(sileht): When we go multi worker, this need to be spread by install_id with an
+                # hashring stored in Redis
+                now = utils.utcnow().timestamp()
+                streams = [
+                    s
+                    for s in reversed(
+                        await self._redis.zrangebyscore("streams", max=now)
```

Comment:
is there any reason to set max and to set it to _now_?

Reply:
When we retries we put a date in the future. We set max to not retrieve them.